### ISSUE TITLE
Diesel migrations breaks diesel with the new feature resolver

### ISSUE THREAD

**Issue by ThouCheese** [diesel-rs/diesel#2753]
### Versions

- rustc --version: `rustc 1.51.0 (2fd73fabe 2021-03-23)`
- **diesel**: `1.4.6`
- **diesel_migrations**: `1.4.0`
- **Postgres**: `psql (PostgreSQL) 13.2`
- **Linux**: 5.11.16 (Arch)

### Feature Flags

- **diesel**: `features = ["postgres"]`
- **diesel_migrations**:  No special features
- **cargo**: resolver = "2"

## Problem Description
When enabling the new resolver in `Cargo.toml` by specifying `resolver = "2"`, compiling `diesel` and `diesel_migrations` fails. In my problem I list postgres as the database, but it fails for `sqlite` and `mysql` too. I tried fiddling with the feature flags of diesel and diesel_migrations, but I couldn't pinpoint the exact problem.

### What are you trying to accomplish?
I'm trying to compile diesel and diesel migrations with the new resolver.

### What is the expected output?
I expect this to compile.

### What is the actual output?
It fails to compile with the following output:
```
error[E0432]: unresolved import `diesel`
  --> /home/me/.cargo/registry/src/github.com-1ecc6299db9ec823/diesel-1.4.6/src/sql_types/mod.rs:37:48
   |
37 | #[derive(Debug, Clone, Copy, Default, QueryId, SqlType)]
   |                                                ^^^^^^^ could not find `pg` in `diesel`
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

### Are you seeing any additional errors?
This error is repeated 26 times for various types in `sql_types/mod.rs`.

### Steps to reproduce
For easy repro, I have created a standalone repository: https://github.com/ThouCheese/diesel-issue. It contains mostly this:

#### `main.rs`
```rust
fn main() {}
```

#### `Cargo.toml`
```toml
[package]
name = "tmp"
version = "0.1.0"
edition = "2018"
resolver = "2"

[dependencies]
diesel = { version = "*", features = ["postgres"] }
diesel_migrations = { version = "*" }
```

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues)  and the [disussion forum](https://github.com/diesel-rs/diesel/discussions)  for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment]
Can you explain why you consider this to be a bug in `diesel` and not as a bug in `cargo`? We do request the corresponding features [here](https://github.com/diesel-rs/diesel/blob/b0c8e99b3e16a81a99c23af1ea90963d2ec7477d/diesel/Cargo.toml#L54).

**Comment by weiznich** [comment]
So I had a short look into this. Seems like cargo is generating the following rustc invocation with `resolver = "2"`:

```
rustc --crate-name diesel ~/.cargo/registry/github.com-1ecc6299db9ec823/diesel-1.4.6/src/lib.rs 
--error-format=json 
--json=diagnostic-rendered-ansi,artifacts 
--crate-type lib 
--emit=dep-info,metadata,link 
-C embed-bitcode=no 
-C debuginfo=2 
-C metadata=8e83eb228db88a74 
-C extra-filename=-8e83eb228db88a74 
--out-dir /tmp/diesel-issue/target/debug/deps 
-L dependency=/tmp/diesel-issue/target/debug/deps 
--extern byteorder=/tmp/diesel-issue/target/debug/deps/libbyteorder-3ba0d986ba1f8d3d.rmeta 
--extern diesel_derives=/tmp/diesel-issue/target/debug/deps/libdiesel_derives-d5f5028a84dbe804.so 
--cap-lints allow 
```

while the following call is generated using the old resolver:

```
rustc --crate-name diesel ~/.cargo/registry/src/github.com-1ecc6299db9ec823/diesel-1.4.6/src/lib.rs 
--error-format=json 
--json=diagnostic-rendered-ansi,artifacts 
--crate-type lib 
--emit=dep-info,metadata,link 
-C embed-bitcode=no 
-C debuginfo=2 
--cfg 'feature="32-column-tables"' 
--cfg 'feature="bitflags"' 
--cfg 'feature="default"' 
--cfg 'feature="postgres"' 
--cfg 'feature="pq-sys"' 
--cfg 'feature="with-deprecated"' 
-C metadata=e4c51cf46b9b40ed 
-C extra-filename=-e4c51cf46b9b40ed 
--out-dir /tmp/diesel-issue/target/debug/deps 
-L dependency=/tmp/diesel-issue/target/debug/deps 
--extern bitflags=/tmp/diesel-issue/target/debug/deps/libbitflags-2a943fc4d3e795d7.rmeta 
--extern byteorder=/tmp/diesel-issue/target/debug/deps/libbyteorder-3ba0d986ba1f8d3d.rmeta 
--extern diesel_derives=/tmp/diesel-issue/target/debug/deps/libdiesel_derives-d5f5028a84dbe804.so 
--extern pq_sys=/tmp/diesel-issue/target/debug/deps/libpq_sys-2f8097271b957b37.rmeta 
--cap-lints allow 
-L native=/usr/lib/x86_64-linux-gnu
```

Notice the missing feature list that is passed to rustc. Without those features it is impossible to get diesel to work. From a short look at the corresponding [documentation](https://doc.rust-lang.org/cargo/reference/resolver.html#feature-resolver-version-2) of `resolver = "2"` it does not seem that we would need to somehow change the way how features are defined in our `Cargo.toml`, so this seems to be a bug in cargo itself.

**Comment by weiznich** [comment]
Opened https://github.com/rust-lang/cargo/issues/9450 for this. I'm closing this bug in favor of the upstream bug, as I cannot see how this is actionable from our side

**Comment by ThouCheese** [comment]
Hey weiznich, thanks for looking into this even though it wasn't a diesel bug!

**Comment by weiznich** [comment]
Let's add an update here: After a bit of discussion with the cargo team in rust-lang/cargo#9450  and https://internals.rust-lang.org/t/concerns-about-making-resolver-2-the-default-in-2021-edition/14684 is is not that clear what the expected behaviour here should be. It seems like that at least more/better diagnostics are required from cargos side. As current workaround it is possible to add the following section to your `Cargo.toml`:
```toml

[build-dependencies]
diesel = {version = "*", features = ["postgres"]}

```

This fixes the compilation error. 

That all written it seems like this should be considered as a bug of similar type like a crate failing to build with `-Z minimal-versions`, which suggests we should eventually fix it in diesel. In fact this specific issue is already fixed on the master branch. Given that + a user needs to explicitly opt into a new feature on cargos side + there is at least a workaround I will likely not provide a proper fix for existing releases. Otherwise if someone provides a fix for the 1.4.x branch I can issue a new patch release as long as the fix does not break anything else.

**Comment by ThouCheese** [comment]
Thanks for the workaround @weiznich. I had kept tabs on the issue you opened in the cargo repository, and it seems a bit sad (to them as well) that this change is not as backwards compatible as they had liked.

**Comment by weiznich** [comment]
This is now addressed with the 1.4.8 release.
